### PR TITLE
Add support for compatibility attributes

### LINKED ISSUE #354: API changes to support compatibility attributes

### PR DESCRIPTION
**Pull request by zerospiel**
* Cluster/Service/Provider Templates now have support for compatibility attributes
* Managed/Management Clusters now have support to report the compatibility attributes
* amends to the templates controller in regards of the API changes
* amends to other parts of the code

Closes #354

### CODE REVIEW

File: api/v1alpha1/clustertemplate_types.go [hunk 1]
```diff
@@ -12,27 +12,120 @@
...
+	"github.com/Masterminds/semver/v3"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // ClusterTemplateSpec defines the desired state of ClusterTemplate
 type ClusterTemplateSpec struct {
-	TemplateSpecCommon `json:",inline"`
+	Helm HelmSpec `json:"helm"`
+	// Compatible K8S version of the cluster set in the SemVer format.
+	KubertenesVersion string `json:"k8sVersion,omitempty"`
```
Comment (Kshatrix) [inline api/v1alpha1/clustertemplate_types.go:30]: for ClusterTemplate it should be the exact version provided by the template

File: api/v1alpha1/clustertemplate_types.go [hunk 2]
```diff
@@ -12,27 +12,120 @@
...
+	// Providers represent required/exposed CAPI providers depending on the template type.
```
Comment (Kshatrix) [inline api/v1alpha1/clustertemplate_types.go:31]:
```suggestion
	// Providers represent required CAPI providers depending on the template type.
```

Comment (Kshatrix) [reply api/v1alpha1/clustertemplate_types.go:31]: this one is semver

File: api/v1alpha1/clustertemplate_types.go [hunk 3]
```diff
@@ -12,27 +12,120 @@
...
+	// Should be set if not present in the Helm chart metadata.
+	Providers ProvidersConstrainted `json:"providers,omitempty"`
 }
 
 // ClusterTemplateStatus defines the observed state of ClusterTemplate
 type ClusterTemplateStatus struct {
+	// Providers represent required/exposed CAPI providers depending on the template type.
+	Providers ProvidersConstrainted `json:"providers,omitempty"`
+
```
Comment (Kshatrix) [inline api/v1alpha1/clustertemplate_types.go:40]: k8sVersion is missing

File: api/v1alpha1/providertemplate_types.go [hunk 4]
```diff
@@ -12,27 +12,120 @@
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // ProviderTemplateSpec defines the desired state of ProviderTemplate
 type ProviderTemplateSpec struct {
-	TemplateSpecCommon `json:",inline"`
+	Helm HelmSpec `json:"helm"`
+	// Compatible CAPI provider version set in the SemVer format.
+	CAPIVersion string `json:"capiVersion,omitempty"`
+	// Providers represent required/exposed CAPI providers depending on the template type.
```
Comment (Kshatrix) [inline api/v1alpha1/providertemplate_types.go:31]:
```suggestion
	// Providers represent exposed CAPI providers.
```

File: api/v1alpha1/providertemplate_types.go [hunk 5]
```diff
@@ -12,27 +12,120 @@
...
+	// Should be set if not present in the Helm chart metadata.
+	Providers ProvidersVersioned `json:"providers,omitempty"`
 }
 
 // ProviderTemplateStatus defines the observed state of ProviderTemplate
 type ProviderTemplateStatus struct {
+	// Providers represent required/exposed CAPI providers depending on the template type.
+	Providers ProvidersVersioned `json:"providers,omitempty"`
+
```
Comment (Kshatrix) [inline api/v1alpha1/providertemplate_types.go:40]: capiVersion is missing

File: api/v1alpha1/servicetemplate_types.go [hunk 6]
```diff
@@ -15,24 +15,75 @@
...
+	// Providers represent required/exposed CAPI providers depending on the template type.
+	// Should be set if not present in the Helm chart metadata.
+	Providers Providers `json:"providers,omitempty"`
 }
 
 // ServiceTemplateStatus defines the observed state of ServiceTemplate
 type ServiceTemplateStatus struct {
+	// Providers represent required/exposed CAPI providers depending on the template type.
+	Providers Providers `json:"providers,omitempty"`
+
```
Comment (Kshatrix) [inline api/v1alpha1/servicetemplate_types.go:37]: k8s version constaint

**Comment by zerospiel** [comment]
Closing in favor of the #422 to simplify the review

**Pull request closed without merging**